### PR TITLE
[FIX] calendar: missing cron compat from sms_calendar

### PR DESCRIPTION
**Pull request by Julien00859**
Fine tunning of c8f031ad77, this commit has the missing code
compatibility for sms_calendar.

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
I would love seeing tests with this fix to ensure it is not silently broken next time.

Many thanks,

**Comment by Julien00859** [comment]
Would had loved to add a test case but there are no tests in this module, I don't feel like being the first one to introduce some, add one commit above and rebase-ff :+1:

**Comment by tde-banana-odoo** [comment]
Mmmh, I will make a task to try to add tests for calendar, calendar_sms and website_calendar reminders. Could be a bit tricky to have something clean.

File: addons/calendar/models/calendar_event.py [hunk 1]
```diff
@@ -688,11 +688,15 @@ def write(self, values):
 
         return True
 
+    def _get_trigger_alarm_types(self):
+        return ['email']
+
     def _setup_alarms(self):
         """ Trigger the cron in the future for every email reminder """
         cron = self.env.ref('calendar.ir_cron_scheduler_alarm')
+        tat = self._get_trigger_alarm_types()
```
Comment (dbeguin) [inline addons/calendar/models/calendar_event.py:697]: "alarm_triggers" ? tat seems a bit too much like the gloomy 'ptav' and 'ptal' to me :D 
I would prefer explicit naming

File: addons/calendar_sms/models/calendar.py [hunk 2]
```diff
@@ -40,33 +43,10 @@ class AlarmManager(models.AbstractModel):
...
-
-        for event in self.env['calendar.event'].browse(events_data):
-            max_delta = events_data[event.id]['max_duration']
-            event_start = fields.Datetime.from_string(event.start)
-            for alert in self.do_check_alarm_for_one_date(event_start, event, max_delta, 0, 'sms', after=last_sms_cron, missing=True):
-                event.browse(alert['event_id'])._do_sms_reminder()
-        return result
+        super()._send_reminder_email()
+        events_ids = self._get_alarmed_events_ids(ttype='sms')
+        events = self.env['calendar.event'].search([('id', 'in', events_ids)])
```
Comment (dbeguin) [inline addons/calendar_sms/models/calendar.py:51]: browse ?
Or even - why not returning browse records in _get_alarmed_events_ids directly ?
And do events.ids when needed ?

**Review by dbeguin — COMMENTED** [review]
Small Review ! 
Bisous!

File: addons/calendar/models/calendar_alarm_manager.py [hunk 3]
```diff
@@ -135,9 +135,7 @@ def do_check_alarm_for_one_date(self, one_date, event, event_maxdelta, in_the_ne
             })
         return result
 
-    @api.model
-    def _send_reminder_email(self, partners=None):
-        # Executed via cron
+    def _get_alarmed_events_ids(self, ttype):
```
Comment (tde-banana-odoo) [inline addons/calendar/models/calendar_alarm_manager.py:138]: _get_event_ids_to_notify ?

File: addons/calendar_sms/models/calendar.py [hunk 4]
```diff
@@ -40,33 +43,10 @@ class AlarmManager(models.AbstractModel):
...
-        cron_interval = cron.interval_number * interval_to_second[cron.interval_type]
-        events_data = self._get_next_potential_limit_alarm('sms', seconds=cron_interval)
-
-        for event in self.env['calendar.event'].browse(events_data):
-            max_delta = events_data[event.id]['max_duration']
-            event_start = fields.Datetime.from_string(event.start)
-            for alert in self.do_check_alarm_for_one_date(event_start, event, max_delta, 0, 'sms', after=last_sms_cron, missing=True):
-                event.browse(alert['event_id'])._do_sms_reminder()
-        return result
+        super()._send_reminder_email()
```
Comment (tde-banana-odoo) [inline addons/calendar_sms/models/calendar.py:49]: Calling super without giving partners ? New python notation or forgot parameter ?

File: addons/calendar_sms/models/calendar.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/calendar_sms/models/calendar.py:51]: +1 _get_alarmed_events_ids could return events, like _get_events_to_notify . Or something like that.

**Review by tde-banana-odoo — COMMENTED** [review]
Posing my peche.

File: addons/calendar_sms/models/calendar.py (hunk 4 above)
Comment (Julien00859) [reply addons/calendar_sms/models/calendar.py:49]: super() with parameters is ugly py2 syntax, super() is shiny beautiful py3 syntax

Comment (tde-banana-odoo) [reply addons/calendar_sms/models/calendar.py:49]: As shiny as new commands for 2many ? NEVER !

#oldmanyellingatclouds

File: addons/calendar_sms/models/calendar.py (hunk 2 above)
Comment (Julien00859) [reply addons/calendar_sms/models/calendar.py:51]: @dbeguin can't use browse as we don't have anything in cache yet, browse doesn't fetch the records from the database, search does.

Comment (tde-banana-odoo) [reply addons/calendar_sms/models/calendar.py:51]: Maybe there is something I did not understand in this code, but you may browse whatever you want, especially those IDs that come from an sql request and therefore already exist. When accessing it in _do_sms_reminder if anything is not in cache, a read will be performed as far as I know.

File: addons/calendar/models/calendar_event.py [hunk 5]
```diff
@@ -688,14 +688,23 @@ def write(self, values):
...
+        alarm_types = self._get_trigger_alarm_types()
+
         for event in self:
-            for alarm in (alarm for alarm in event.alarm_ids if alarm.alarm_type == 'email'):
-                cron._trigger(at=event.start-timedelta(minutes=alarm.duration_minutes))
+            for alarm in (alarm for alarm in event.alarm_ids if alarm.alarm_type in alarm_types):
+                at = event.start - timedelta(minutes=alarm.duration_minutes)
+                if at > cron.lastcall:
+                    # Don't trigger for past alarms, they would be skipped by design
+                    cron._trigger(at=at)
```
Comment (Julien00859) [inline addons/calendar/models/calendar_event.py:705]: We don't want to actively send reminder for any past alarm. Imagine should schedule at 8 o'clock a meeting at 16 o'clock with a 1h reminder. At 15h30 you decide to rename the meeting (or anything else), you don't want to resend an alarm. If you schedule at 8 a meeting at 9 with a 2h reminder the attendees receive an invitation email already so the reminder is not really needed.

File: addons/calendar_sms/models/calendar.py (hunk 2 above)
Comment (Julien00859) [reply addons/calendar_sms/models/calendar.py]: My bad... both of you are right, one day I'll be able to use that ORM I help coding...

File: addons/calendar/models/calendar_event.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/calendar/models/calendar_event.py:705]: It is waaaaay to early to understand all that computation.

:coffee:

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**